python_benchmarks/core.py

```diff
@@ -0,0 +1,202 @@
+
+        Class members:
+            self.prof: torch.profiler instance used as by the custom torchprofile_timer for the current benchmark
+            self.benchmark: Underlying pytest-benchmark fixture with timer modified to use torchprofile_timer
+            self.current_time: Global montonic clock incremented based on elapsed CUDA time
+        """
+        # Initialize a Torch Profiler object
+        self.prof = profile(activities=[ProfilerActivity.CUDA, ProfilerActivity.CPU])
+
+        # Modify the default timer.
+        benchmark_fixture._timer = self.torchprofile_timer
+        # Externally set the precision to avoid timer calibration. Since the timer uses CUDA times,
+        # calibration using subsequent timer calls produces invalid results.
+        # https://github.com/ionelmc/pytest-benchmark/blob/728752d2976ef53fde7e40beb3e55f09cf4d4736/src/pytest_benchmark/timers.py#L15
+        benchmark_fixture._precisions[benchmark_fixture._timer] = precision
+
+        self.benchmark = benchmark_fixture
+
+        # Global montonic clock
+        self.current_time = 0.0
+
+    def __call__(self, function_to_benchmark: Callable, *args, **kwargs):
+        return self.benchmark(function_to_benchmark, *args, **kwargs)
+
+    def __getattr__(self, attr):
+        if attr not in self.__dict__:
+            return getattr(self.benchmark, attr)
+        return super().__getattr__(attr)
+
+    def torchprofile_timer(self) -> float:
+        """
+        Custom torchprofiler-based timer used by pytest-benchmark.
+        At every timer call, the profiler is stopped to compute the elapsed CUDA time
+        and the global clock is incremented. The profiler is restarted before returning to continue tracing.
+
+        Returns:
+            self.current_time: Global monotonic clock variable
+        """
+        try:
+            self.prof.stop()
+            prof_averages = self.prof.key_averages()
+            elapsed_cuda_time = self._get_kernel_time(prof_averages)
+            self._increment_global_time(elapsed_cuda_time)
+            self.prof.start()
+        except:
+            self.prof.start()
+        return self.current_time
+
+    def _get_kernel_time(self, prof_averages) -> float:
+        """
+        Returns:
+            time_value: Elapsed CUDA time in seconds.
+        """
+        elapsed_cuda_time = sum([event.self_cuda_time_total
+            for event in prof_averages 
+            if event.device_type == DeviceType.CUDA])/ 1e6
+        return elapsed_cuda_time
+
+    def _increment_global_time(self, elapsed_time: float) -> None:
+        self.current_time += elapsed_time
+
+    def cleanup(self) -> None:
+        """
+        Stops a running torchprofiler instance if found.
+        """
+        try:
+            self.prof.stop()
+        except:
+            pass
+
+    def compute_metrics(
+        self, inputs: Union[torch.Tensor, List], outputs: Union[torch.Tensor, List]
+    ) -> None:
```

Comment:
I'd expect this method to return the metrics rather than store the metrics in fields. 

With
```
nvf_benchmark.compute_metrics(...)
metrics = self.foo
````
it's unclear whether `self.foo` is computed by `compute_metrics` or computed before and just accessed here. 

Is there a strong reason that you chose the latter?

Reply:
I did not choose to return the metrics since they are not returned by `run_benchmark` or used subsequently in the function. The metrics we compute may also change, so I would avoid returning those values. Maybe a better name for this function would be `set_metrics`.